compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -0,0 +1,951 @@
+        assert(name.isTermName, "type trees should not reach here")
+        cases(expr.tpe, thisV, klass, expr)
+
+      case NewExpr(tref, New(tpt), ctor, argss) =>
+        // check args
+        val errors = evalArgs(argss.flatten, thisV, klass)
+
+        val cls = tref.classSymbol.asClass
+        val res = outerValue(tref, thisV, klass, tpt)
+        (res ++ errors).instantiate(cls, ctor, expr)(using heap, ctx, trace.add(expr))
+
+      case Call(ref, argss) =>
+        // check args
+        val errors = evalArgs(argss.flatten, thisV, klass)
+
+        val trace2: Trace = trace.add(expr)
+
+        ref match
+        case Select(supert: Super, _) =>
+          val SuperType(thisTp, superTp) = supert.tpe
+          val thisValue2 = resolveThis(thisTp.classSymbol.asClass, thisV, klass, ref)
+          Result(thisValue2, errors).call(ref.symbol, superTp, expr)(using heap, ctx, trace2)
+
+        case Select(qual, _) =>
+          val res = eval(qual, thisV, klass) ++ errors
+          res.call(ref.symbol, superType = NoType, source = expr)(using heap, ctx, trace2)
+
+        case id: Ident =>
+          id.tpe match
+          case TermRef(NoPrefix, _) =>
+            // resolve this for the local method
+            val enclosingClass = id.symbol.owner.enclosingClass.asClass
+            val thisValue2 = resolveThis(enclosingClass, thisV, klass, id)
+            thisValue2 match
+            case Hot => Result(Hot, errors)
+            case _ =>
+              val rhs = id.symbol.defTree.asInstanceOf[DefDef].rhs
+              eval(rhs, thisValue2, enclosingClass, cacheResult = true)(using heap, ctx, trace2)
+          case TermRef(prefix, _) =>
+            val res = cases(prefix, thisV, klass, id) ++ errors
+            res.call(id.symbol, superType = NoType, source = expr)(using heap, ctx, trace2)
+
+      case Select(qualifier, name) =>
+        eval(qualifier, thisV, klass).select(expr.symbol, expr)
+
+      case _: This =>
+        cases(expr.tpe, thisV, klass, expr)
+
+      case Literal(_) =>
+        Result(Hot, noErrors)
+
+      case Typed(expr, tpt) =>
+        if (tpt.tpe.hasAnnotation(defn.UncheckedAnnot)) Result(Hot, noErrors)
+        else eval(expr, thisV, klass) ++ checkTermUsage(tpt, thisV, klass)
+
+      case NamedArg(name, arg) =>
+        eval(arg, thisV, klass)
+
+      case Assign(lhs, rhs) =>
+        lhs match
+        case Select(qual, _) =>
+          val res = eval(qual, thisV, klass)
+          eval(rhs, thisV, klass).ensureHot("May only assign initialized value", rhs) ++ res.errors
+        case id: Ident =>
+          eval(rhs, thisV, klass).ensureHot("May only assign initialized value", rhs)
+
+      case closureDef(ddef) =>
+        thisV match
+        case obj: (ThisRef | Warm) =>
+          val value = Fun(ddef.rhs, obj, klass)
+          Result(value, Nil)
+        case _ =>
+          // The reason is that we never evaluate an expression if `thisV` is
```

Comment:
Can you explain why we don't evaluate an expression when thisV is cold?

Reply:
That's part of the design of the system -- when we call a method on a cold value, the system reports a warning. That's why this branch cannot be reached.